### DISCUSSION TITLE
How to: deep nested relationship api calls

### DISCUSSION THREAD

**Discussion by nokia6290** [pocketbase/pocketbase#4206] [Q&A] [answered]
Hi,

I'm exploring pocketbase as a potential backend for my side project.
Let's assume I'm building a gym app tracking ->
Collections: **workout**, **exercise_session**, **session_set**. Workout_id will have 1 to many **exercise_session** and **exercise_session** will have 1 to many **session_set** relationships.

My question is so how do I make a single api call when uploading a workout with let's say 10 exercise_session, and 10 session_set in each. It's a single workout object, but how to build the relationships between the tables in one API call?

I've read that making multiple API calls might work, but that sounds anti-pattern.

Can someone guide me in how to achieve this?

Thank you

**Comment by ganigeorgiev** [comment]
Please see https://pocketbase.io/docs/working-with-relations/.

> My question is so how do I make a single api call when uploading a workout with let's say 10 exercise_session, and 10 session_set in each. It's a single workout object, but how to build the relationships between the tables in one API call?

You can't, because bulk/transactional web APIs are currently not supported. You need to create all relations beforehand to assign their id(s) to the main/parent record.

**Reply by nokia6290** [reply]
Thank you for replying.

Creating first objects would require me to send in some cases 20-50 of API calls which would definitely not work well..

What about extending the framework and creating custom endpoint with GO? Would it then be feasible to send a single API as a json object and the logic would in theory handle the relationships between objects?

**Reply by ganigeorgiev** [reply]
You can create a custom route with Go or JS, but based on your limited description I personally will use the existing APIs and create the relations separately, because this will also allow to persist "partially" the user entered data so that they can continue at later stage from where they left or depending on your UI and how the form is structured you may want to have separate screens for each entry (for example first creating the basic "workout" record, then entering one by one the details about "exercise_session", etc.).

**Reply by nokia6290** [reply]
The partial persistence is doable, but actually until a workout is finished any data logged might be deleted so there is a scenario where a user logs 10-20 exercises, client side does 20 api calls just to create relationships and then during the workout session the values or order change and we have to send additional API calls. This can grow exponentially.
A local db or state can hold the partial persistent until a workout is logged do server.

VS
having 1 single API call to log a finished workout.

Though I will probably test both scenarios of logging partially data to server and extending Go framework with a custom route.

@ganigeorgiev may I ask, is the bulk transaction on the pipeline to be developed this year?

**Comment by jonshipman** [comment, marked as answer]
To add, you can run `$security.randomString(15);` in a Pocketbase pb_hooks file to generate an id beforehand. In theory, you'd create a custom POST route, parse the body, and just import the records as you deem acceptable. Of note: I do recommend using either the query builder or newQuery functions as record operations will prevent you from adding relationships that don't exist.

So you can make a `/custom-api/worker` route and send the worker payload in the body:

```
routerAdd("POST", "/custom-api/workout", (c) => {
  try {
    const body = new DynamicModel({
      worker: {},
    });
    c.bind(body);

    const { worker } = body;

    if (worker.exercise_session) {
      worker.exercise_session = worker.exercise_session.map((session) => {
        session.id = $security.randomString(15);
        $app.dao().db().insert("exercise_session", session).execute();

        if (session.session_set) {
          session.session_set = session.session_set.map((set) => {
            set.id = $security.randomString(15);
            $app.dao().db().insert("session_set", set).execute();

            return set.id;
          });
        }

        return session.id;
      });
    }

    $app.dao().db().insert("workout", worker).execute();
  } catch (error) {
    throw new ApiError(500, error.message);
  }
});
```

YMMV depending on other values in the body. Note: I haven't tested the code - you may need to stringify the arrays before insertion.

**Reply by nokia6290** [reply]
Thank you.

Will try this approach as well. 
Can you tell me more about the **query builder or newQuery functions** . Where can I read more about it?

edit: found the query builder documentation https://pocketbase.io/docs/go-database/#query-builder

The query builder at first glance looks like it does allow big customization

users := []struct {
    Id    string `db:"id" json:"id"`
    Email bool   `db:"email" json:"email"`
}{}

app.Dao().DB().
    Select("id", "email").
    From("users").
    AndWhere(dbx.Like("email", "example.com")).
    Limit(100).
    OrderBy("created ASC").
    All(&users)

**Reply by nokia6290** [reply]
Would you recommend adding the custom API route directly into pocketbase main.go frame or create a separate implementation for it?

Adding directly would be quicker, but updates will become a hassle. While separate implementation will require more careful planning but easier to update/maintain I guess
